Ap add all

BabyBoySnow opened the pull request:
After creating your pull request, tick these boxes if they are applicable to you.

- [x] I have tested my changes against the `fae` branch (the latest developmental version), and this pull request is targeting that branch as a base
- [x] I have tested my changes on Python 3.8 or higher
- [x] I have ensured my code is formatted using [Black](https://github.com/psf/black)

----

### Description
Adds a new sub command to autoplaylist to save the entire current queue. 


### Related issues (if applicable)

BabyBoySnow committed:
add sub option save all

Create new sub command to add the entire queue to the autoplaylist.
Also felt that having the 's out side of the `` looked kinda weird...

BabyBoySnow committed:
bro really?

itsTheFae commented on musicbot/bot.py:
```diff
@@ -3129,6 +3132,41 @@ def _get_url() -> str:
                 )
             return url
 
+        if option in ["+all", "add_all"]:
```
This will work, but as you mentioned wanting to use `!autoplaylist [+|add] all` in discord, for that you would want to inspect the value of `opt_url` directly.  If no "url" command parameter is given `opt_url` is simply an empty string as per its default value in the function definition.   

Since this command currently does not allow users to add songs which are not URLs, reserving the word "all" wouldn't be an issue.  If later we decide song titles should be allowed through this function, this command would need some refactor to allow that anyways. 

So dealers choice here, I'm good with this if you are.

BabyBoySnow replied:
I've taken the route of reserving "all" as a special case of opt_url.

itsTheFae commented on musicbot/bot.py:
```diff
@@ -3129,6 +3132,41 @@ def _get_url() -> str:
+            added_songs = set()
+            for e in player.playlist.entries:
+                url = e.url
+                self._do_song_blocklist_check(url)
```
We can probably avoid doing blocklist checking for this case.  In theory, the songs that have made it into the queue have already been checked.  (Just not 100% sure about playlist entries, but they should be too.)

itsTheFae commented on musicbot/bot.py:
```diff
@@ -3129,6 +3132,41 @@ def _get_url() -> str:
+                self._do_song_blocklist_check(url)
+                if e.url not in self.server_data[guild.id].autoplaylist:
+                    await self.server_data[guild.id].autoplaylist.add_track(url)
+                    added_songs.add(url)
```
Just a nit-pick more than anything, but why not use `e.url` in place of all the `url` in this loop here?

BabyBoySnow replied:
copy paste :p

itsTheFae commented on musicbot/bot.py:
```diff
@@ -3096,12 +3098,13 @@ async def cmd_autoplaylist(
         """
         Usage:
-            {command_prefix}autoplaylist [ + | - | add | remove] [url]
+            {command_prefix}autoplaylist [ + | +all | - | add | add_all | remove] [url]
```
May be worth adding a second line since the `+all|add_all` sub-commands do not take the `[url]` command parameter

BabyBoySnow committed:
pr feedback stuf

reserve "all" as a special case of opt_url to allow spaces.
just use e.url instead.
Remove the blocksong check, if it's in queue it should've already passed that.
Add second docstring to show that + all and add all don't need a url.

itsTheFae committed:
Enforce add/+ for all, adjust space in help text
